# Тести для малих вибірок та пропорцій



**Завдання 1**. E-commerce компанія після редизайну сайту підозрює, що **середній час до покупки (time-to-purchase)** користувачів **збільшився**.
Історично середній час від першого заходу на сайт до покупки становив $\mu_0 = 123.7$ хвилин.

Команда аналітиків випадково обрала дані **7 покупців після редизайну**:
`128, 135, 121, 142, 126, 151, 123`

З допомогою наявних даних зʼясуйте, чи збільшився середній час покупки після редизайну.

Для цього виконайте наступні 6 кроків. Правильне виконання кожного з кроків оцінюється в 1 бал.

1. Запишіть нульову та альтернативну гіпотези і визначте тип тесту.
2. Обчисліть вибіркові статистики: $\bar x$, $s$, $n$.
3. Оберіть тип тесту та виконайте його (знайдіть критичне значення тесту, статистику тесту та р-значення) будь-яким пасуючим способом, щоб перевірити гіпотезу на рівні значущості $\alpha = 0.10$.
4. Прийміть рішення, чи відхиляєте ви гіпотезу $H_0$ використовуючи p-value.
5. Напишіть висновок: чи справді редизайн сайту подовжив час до покупки?
6. Чи зміниться ваше рішення при зміні рівня значущості на $\alpha = 0.05$.


In [1]:
import numpy as np
from scipy import stats

time = np.array([128, 135, 121, 142, 126, 151, 123])
mu0 = 123.7   
alpha = 0.1
alpha2 = 0.05

# Крок 1. Обчислимо вибіркову статистику
x_bar = np.mean(time)
s = np.std(time, ddof=1)   
n = len(time)

# Крок 2. t-статистика
t_stat = (x_bar - mu0) / (s / np.sqrt(n))

# Крок 3. Критичне значення для правостороннього тесту
df = n - 1
t_crit = stats.t.ppf(1 - alpha, df)

# Крок 4. p-value для правостороннього тесту
p_value = 1 - stats.t.cdf(t_stat, df)

print(f"Вибіркове середнє: {x_bar:.2f}")
print(f"Вибіркове стандартне відхилення: {s:.2f}")
print(f"t-статистика: {t_stat:.3f}")
print(f"Критичне значення (df={df}): {t_crit:.3f}")
print(f"p-value: {p_value:.4f}")

# Крок 5. Висновок
if t_stat > t_crit:
    print(f"Відхиляємо H0: середній час більший за {mu0} на рівні значущості {alpha}")
else:
    print("Не відхиляємо H0: доказів недостатньо")



Вибіркове середнє: 132.29
Вибіркове стандартне відхилення: 10.98
t-статистика: 2.069
Критичне значення (df=6): 1.440
p-value: 0.0420
Відхиляємо H0: середній час більший за 123.7 на рівні значущості 0.1


# Рівень значущості = 0.05

In [2]:
# Крок 1. Обчислимо вибіркову статистику
x_bar = np.mean(time)
s = np.std(time, ddof=1)   # стандартне відхилення з n-1
n = len(time)

# Крок 2. t-статистика
t_stat = (x_bar - mu0) / (s / np.sqrt(n))

# Крок 3. Критичне значення для правостороннього тесту
df = n - 1
t_crit = stats.t.ppf(1 - alpha2, df)

# Крок 4. p-value для правостороннього тесту
p_value = 1 - stats.t.cdf(t_stat, df)

print(f"Вибіркове середнє: {x_bar:.2f}")
print(f"Вибіркове стандартне відхилення: {s:.2f}")
print(f"t-статистика: {t_stat:.3f}")
print(f"Критичне значення (df={df}): {t_crit:.3f}")
print(f"p-value: {p_value:.4f}")

# Крок 5. Висновок
if t_stat > t_crit:
    print(f"Відхиляємо H0: середній час більший за {mu0} на рівні значущості {alpha2}")
else:
    print("Не відхиляємо H0: доказів недостатньо")


Вибіркове середнє: 132.29
Вибіркове стандартне відхилення: 10.98
t-статистика: 2.069
Критичне значення (df=6): 1.943
p-value: 0.0420
Відхиляємо H0: середній час більший за 123.7 на рівні значущості 0.05


Висновок: Отже редизайн сайту збільшив середній час перед покупкою. 
При $\alpha = 0.05$ нам тяжче було відкинути нульову гіпотезу оскільки 2.069 наближено до критичного значпення 1,943. При $\alpha = 0.1$0 ми бачимо значну різницю між t-статистика 2.069 та критичним значенням 1.440, тому збільшою впевненістю можна відкинути початкову гіпотезу.


**Завдання 2.**

До спеціальної рекламної кампанії **23%** дорослих упізнавали логотип компанії. Після завершення кампанії відділ маркетингу провів опитування: з **1200** випадково відібраних дорослих **311** упізнали логотип.

Перевірте на рівні значущості **3%** ($\alpha=0.03$), чи дають ці дані достатні підстави стверджувати, що **тепер більше ніж 23%** дорослих упізнають логотип компанії. Для розвʼязку використовуйте бібліотеку `statsmodels`.

Зробіть висновок, чи зросла впізнаваність логотипу.

Додатково, обчисліть довірчий інтревал на заданому рівні значущості і проінтерпретуйте текстом - як він додатково пояснює прийняте нами рішення?

In [6]:
import statsmodels.api as sm
from statsmodels.stats.proportion import proportions_ztest, proportion_confint

count = 311   # кількість "успіхів"
nobs = 1200    # обсяг вибірки
p0 = 0.23      # гіпотетична пропорція (H0)

stat, pval = proportions_ztest(count, nobs, p0, alternative='larger')
confint = proportion_confint(count, nobs, alpha=0.03, method='normal')


print(f"Z-статистика = {stat:.3f}")
print(f"p-value = {pval:.4f}")
print(f"97% довірчий інтервал: {confint}")

if pval < alpha:
    print("Висновок:") 
    print(f"p-value = {pval:.4f} < α = {alpha}, тому відхиляємо H₀.")
    print("Маємо достатньо доказів, щоб стверджувати, що впізнаваність логотипу зросла понад 23%.")
    print(f"97% довірчий інтервал: ({confint[0]:.3f}, {confint[1]:.3f})")
    print(f"Це означає, що частка дорослих, які впізнають логотип, лежить між {confint[0]*100:.1f}% та {confint[1]*100:.1f}%")
else:
    print("Висновок:") 
    print(f"p-value = {pval:.4f} ≥ α = {alpha}, тому НЕ відхиляємо H₀.")
    print("Достатніх доказів про збільшення впізнаваності логотипу немає.")
    print(f"97% довірчий інтервал: ({confint[0]:.3f}, {confint[1]:.3f})")

Z-статистика = 2.306
p-value = 0.0106
97% довірчий інтервал: (0.23171700302179205, 0.28661633031154127)
Висновок:
p-value = 0.0106 < α = 0.1, тому відхиляємо H₀.
Маємо достатньо доказів, щоб стверджувати, що впізнаваність логотипу зросла понад 23%.
97% довірчий інтервал: (0.232, 0.287)
Це означає, що частка дорослих, які впізнають логотип, лежить між 23.2% та 28.7%
